components/kubelet/pod/component.go

```diff
@@ -168,17 +167,23 @@ func (d *Data) getReason() string {
 			return fmt.Sprintf("connection error to node %q -- %v", d.NodeName, d.err)
 		}
 
-		return fmt.Sprintf("failed to list pods from kubelet read-only port -- %v", d.err)
+		return fmt.Sprintf("list pods from kubelet read-only port failed %d time(s) (%v)", failedCount, d.err)
 	}
 
 	return fmt.Sprintf("total %d pods (node %s)", len(d.Pods), d.NodeName)
 }
 
-func (d *Data) getHealth(ignoreConnErr bool) (string, bool) {
+func (d *Data) getHealth(ignoreConnErr bool, failedCount int, failedCountThreshold int) (string, bool) {
 	healthy := d == nil || d.err == nil
+
 	if d != nil && d.err != nil && d.connErr && ignoreConnErr {
 		healthy = true
 	}
+
+	if d != nil && failedCount >= failedCountThreshold {
+		healthy = false
+	}
```

Comment:
we should move the logical to checkonce function... the data is just used to trace the last state, but only checkonce function is responsible for determine whether the kubelet is healthy...

Reply:
@eahydra Done